Sources/FoundationEssentials/String/String+Internals.swift

```diff
@@ -52,5 +60,194 @@ extension String {
+            return _decomposed(.hfsPlus, into: newBuffer, nullTerminated: true)
+        }
+        
+        return result != nil
+    }
+    #endif
+    
+    package func withFileSystemRepresentation<R>(_ block: (UnsafePointer<CChar>?) throws -> R) rethrows -> R {
+        #if canImport(Darwin)
+        try withUnsafeTemporaryAllocation(of: CChar.self, capacity: Int(MAXPATHLEN)) { buffer in
+            guard _fileSystemRepresentation(into: buffer) else {
+                return try block(nil)
+            }
+            return try block(buffer.baseAddress!)
+        }
+        #else
+        try self.withCString {
+            try block($0)
+        }
+        #endif
+    }
+}
 
+extension UnsafeBufferPointer {
+    private enum DecompositionError : Error {
+        case insufficientSpace
+        case illegalScalar
+        case decodingError
+    }
+    
+    fileprivate func _decomposedRebinding<T: UnicodeCodec, InputElement>(_ type: String._NormalizationType, as codec: T.Type, into buffer: UnsafeMutableBufferPointer<InputElement>, nullTerminated: Bool = false) throws -> Int {
+        try self.withMemoryRebound(to: T.CodeUnit.self) { reboundSelf in
+            try buffer.withMemoryRebound(to: Unicode.UTF8.CodeUnit.self) { reboundBuffer in
+                try reboundSelf._decomposed(type, as: codec, into: reboundBuffer, nullTerminated: nullTerminated)
+            }
+        }
+    }
+    
+    fileprivate func _decomposed<T: UnicodeCodec>(_ type: String._NormalizationType, as codec: T.Type, into buffer: UnsafeMutableBufferPointer<UInt8>, nullTerminated: Bool = false) throws -> Int where Element == T.CodeUnit {
+        let scalarSet = BuiltInUnicodeScalarSet(type: type.setType)
+        var bufferIdx = 0
+        let bufferLength = buffer.count
+        var sortBuffer: [UnicodeScalar] = []
+        var seenNullIdx: Int? = nil
+        var decoder = T()
+        var iterator = self.makeIterator()
+        
+        func appendOutput(_ values: some Sequence<UInt8>) throws {
+            let bufferPortion = UnsafeMutableBufferPointer(start: buffer.baseAddress!.advanced(by: bufferIdx), count: bufferLength - bufferIdx)
+            var (leftOver, idx) = bufferPortion.initialize(from: values)
+            bufferIdx += idx
+            if bufferIdx == bufferLength && leftOver.next() != nil {
+                throw DecompositionError.insufficientSpace
+            }
+        }
+        
+        func appendOutput(_ value: UInt8) throws {
+            guard bufferIdx < bufferLength else {
+                throw DecompositionError.insufficientSpace
+            }
+            buffer.initializeElement(at: bufferIdx, to: value)
+            bufferIdx += 1
+        }
+        
+        func encodedScalar(_ scalar: UnicodeScalar) throws -> some Collection<UInt8> {
+            guard let encoded = UTF8.encode(scalar) else {
+                throw DecompositionError.illegalScalar
```

Comment:
This would be the case of an unpaired surrogate character (which has no valid UTF-8 representation but is a valid unicode scalar). If you have an `NSString` that calls to this decomposition with its UTF-16 buffer, it will successfully decode an unpaired surrogate to a unicode scalar but will then fail to encode it back to UTF-8